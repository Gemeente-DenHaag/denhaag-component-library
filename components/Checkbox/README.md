# Checkbox

![npm (scoped)](https://img.shields.io/npm/v/@gemeente-denhaag/checkbox?logo=npm&style=flat-square)
![GitHub Workflow Status (branch)](https://img.shields.io/github/workflow/status/nl-design-system/denhaag/Build%20and%20deploy%20Storybook%20to%20Azure%20Web%20App/main?logo=github&style=flat-square)

Checkboxes are usually used to provide a way for users to make a range of selections (zero, one or more). They can also be used to tell users that they agree to specific terms.

## When to use

Checkboxes are used for multiple choices, not for mutually exclusive choices. Each checkbox works independently from other checkboxes in the list, therefore checking an additional box does not affect any other selections.

### Forms

Can be used in forms on a full page or in modals.

### Terms and conditions

Turning the checkbox input on or off can indicate whether you agree to the terms.

## Alternatives and related components

Use radio buttons to display a list of options where users can only make one choice

[Navigate to Radio](../?path=/docs/components-input-radio--default)

## Anatomy

The checkbox consists of:

1. Label (optional): indicates what should be selected below.
2. Checkbox input: a checkbox input that indicates the correct status. By default it is not active.
3. Checkbox label: describes the information you can select or deselect.
4. Helper text (optional): displays more information about the checkboxes

## (Interactive) states

The checkbox contains the states normal, hover, disabled, error and focus.

## Design properties

### Typography

- Label: TheSans/md/700
- Checkbox label: TheSans/md/400
- Helper text: TheSans/s/400

### Colors

- Label: text color Blue/5
- Checkbox input inactive: border color Grey/4, background-color white
- Checkbox input active: border color Blue/3, background-color Blue/3, checkmark icon white
- Checkbox label: text color Grey/4
- Helper text: text color Grey/4

### Interactive states

- Disabled: container border color Grey/2, checkbox label text color Grey/2
- Error: container border color Red/3, checkbox label text color Grey/4
- Focus: container border color Ocher/5

### Structure

- Checkbox: height and width 20px, border 1px
- Checkbox focus: border 2px, dash 5,5
- Label: margin-bottom 8px
- Checkbox label: padding-left 8px
- Helper text: margin-left 16px, margin-top 4px

## Accessibility

Screen readers automatically report the status of the check box.

- [technical accessibility requirements]

### Labeling

[technical requirements, such as: label and labelHidden]

### External links

[technical requirements]

### Keyboard support

- Move focus to each checkbox using the tab key (or shift + tab when tabbing backwards)
- Activate or deactivate checkboxes with the space key

## Content guidelines

### Checkbox labels

Checkbox labels must:

- Start with a capital letter
- Not end in punctuation if it’s a single sentence, word, or a fragment
- In case the checkbox asks to agree to the terms and conditions, use the first person

## Best practices

### Do's

Checkboxes must:

- Work independently of each other: selecting a checkbox may not change the selection status of another checkbox in the list.
- Be positively boxed: for example “Enable notifications” instead of “Disable notifications”.
- Always have a label handy when you use it to turn a setting on or off.
- Use label tags as click targets.
- Be arranged in a logical order, whether alphabetical, numeric, time-based or some other clear system.
- Link to more information or have a subtitle for more explanation. Don't rely on tooltips to explain a checkbox.
- If they are too long, run over the second line. This is preferable than truncation.
- If they are too long, they will be aligned left
- If they are too long, let the text flow under the checkbox so that the control and label are aligned at the top.

### Don'ts

Checkboxes must:

- Not cut with an omit task if the label is too long and run over 2 lines.

### Buttons versus links

Buttons are mainly used for actions such as "Add", "Close", "Cancel" or "Save". Regular buttons similar to links are used for less important or less frequently used actions, such as "View details".

Links are mainly used for navigation and usually appear in or immediately after a sentence.

The HTML displayed for the Button and Link components has meaning. Using these components intentionally and consistently results in:

- a more inclusive experience for assistive technology users
- a more coherent visual experience for sighted users
- products that are easier to maintain to scale

## References

https://www.nngroup.com/articles/checkboxes-vs-radio-buttons/ </br>
https://uxplanet.org/checkbox-and-toggle-in-forms-f0de6086ac41 </br>
