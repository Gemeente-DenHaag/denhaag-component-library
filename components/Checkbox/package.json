{
  "name": "@gemeente-denhaag/checkbox",
  "version": "0.2.3-alpha.160",
  "description": "A Checkbox component",
  "bugs": "https://github.com/nl-design-system/denhaag/issues",
  "repository": {
    "type": "git",
    "url": "https://github.com/nl-design-system/denhaag.git",
    "directory": "components/Checkbox"
  },
  "license": "EUPL-1.2",
  "author": "Municipality of The Hague",
  "exports": "./dist/index.js",
  "main": "dist/index.js",
  "module": "dist/index.js",
  "types": "dist/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "rollup -c ../../rollup.config.js",
    "clean": "yarn rimraf dist tsconfig.tsbuildinfo"
  },
  "dependencies": {
    "@gemeente-denhaag/icons": "0.2.3-alpha.160"
  },
  "devDependencies": {
    "@gemeente-denhaag/form-field": "0.1.1-alpha.36",
    "@gemeente-denhaag/formcontrollabel": "0.2.3-alpha.160"
  },
  "peerDependencies": {
    "react": "^17.0.0"
  },
  "gitHead": "dcf72a9b79266c1ebede35aff4a02dd9121a980f"
}
