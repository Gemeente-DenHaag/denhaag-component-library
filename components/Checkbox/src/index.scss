@import "~@utrecht/components/dist/checkbox/css";

.utrecht-checkbox--custom {
  border-radius: var(--utrecht-checkbox-border-radius);
}

.utrecht-checkbox--disabled {
  background-color: var(--utrecht-checkbox-disabled-background-color);
}
