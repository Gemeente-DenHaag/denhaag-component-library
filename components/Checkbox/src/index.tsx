import React, { InputHTMLAttributes } from 'react';
import clsx from 'clsx';
import { CheckedIcon } from '@gemeente-denhaag/icons';

import './index.scss';

export interface CheckboxProps extends InputHTMLAttributes<HTMLInputElement> {
  /**
   * Displays the checkbox red.
   */
  error?: boolean;
  /**
   * Disables Checkbox
   */
  disabled?: boolean;
  /**
   * Callback fired when the state is changed.
   */
  onChange?: (event: React.ChangeEvent<unknown>) => void;
  /**
   * Overwrite the Icon shown when in the checked state.
   */
  checkedIcon?: React.ReactNode;
}

/**
 * Checkboxes allow the user to select one or more items from a set.
 */
export const Checkbox: React.FC<CheckboxProps> = ({
  className,
  error = false,
  checked = false,
  checkedIcon = <CheckedIcon />,
  onChange,
  ...props
}: CheckboxProps) => {
  const [isChecked, setChecked] = React.useState(checked);
  const classes = clsx('denhaag-checkbox', { 'denhaag-checkbox--error': error }, className);

  return (
    <span className={classes}>
      <input
        type="checkbox"
        className="denhaag-checkbox__input"
        checked={isChecked}
        onChange={(event) => {
          setChecked(!isChecked);
          if (onChange) {
            onChange(event);
          }
        }}
        {...props}
      />
      <span className="denhaag-checkbox__icon">{checkedIcon}</span>
    </span>
  );
};

export default Checkbox;
