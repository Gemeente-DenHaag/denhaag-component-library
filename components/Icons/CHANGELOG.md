# @gemeente-denhaag/icons

## 1.0.0

### Major Changes

- 281fa34: Added wallet icon
