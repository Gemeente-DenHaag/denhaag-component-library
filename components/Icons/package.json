{
  "name": "@gemeente-denhaag/icons",
  "version": "0.2.3-alpha.84",
  "description": "The icon set of the Municipality of The Hague",
  "bugs": "https://github.com/nl-design-system/denhaag/issues",
  "repository": {
    "type": "git",
    "url": "https://github.com/nl-design-system/denhaag.git",
    "directory": "components/Icons"
  },
  "license": "EUPL-1.2",
  "author": "Municipality of The Hague",
  "exports": "./dist/index.js",
  "main": "dist/index.js",
  "module": "dist/index.js",
  "types": "dist/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "rollup -c ../../rollup.config.js",
    "clean": "yarn rimraf dist tsconfig.tsbuildinfo"
  },
  "dependencies": {
    "@gemeente-denhaag/baseprops": "0.2.3-alpha.84",
    "@material-ui/core": "4.12.3"
  },
  "peerDependencies": {
    "react": "^17.0.0"
  }
}
