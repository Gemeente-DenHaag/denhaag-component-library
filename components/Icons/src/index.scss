.denhaag-icon {
  aspect-ratio: 1;
  display: inline-block;
  fill: currentcolor;
  flex-shrink: 0;
  font-size: var(--denhaag-icon-size, var(--denhaag-space-xl));
  height: 1em;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  width: 1em;
}
