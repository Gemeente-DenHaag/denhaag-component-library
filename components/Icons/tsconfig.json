{
  "extends": "../../tsconfig.build.json",
  "compilerOptions": {
    "rootDir": "src",
    "outDir": "dist",
    "typeRoots": ["./src/@types"],
    "composite": true
  },
  "include": ["src/**/*"],
  "exclude": ["dist", "src/**/*.stories.tsx"]
}
