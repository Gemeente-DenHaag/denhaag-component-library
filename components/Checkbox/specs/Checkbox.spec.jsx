import * as React from 'react';
import Checkbox from '../dist';
import { HideIcon } from '../../Icons';

describe(`Checkbox tests`, () => {
  it('can be clicked', () => {
    const props = {
      onChange: () => {
        console.log('Changed');
      },
    };

    const spy = cy.spy(props, 'onChange').as('Input Change');

    cy.mount(<Checkbox onChange={props.onChange}></Checkbox>);

    cy.get('.denhaag-checkbox')
      .click()
      .then(() => expect(spy).to.have.been.calledOnce);

    cy.get('.denhaag-checkbox__input').should('be.checked');
    cy.get('.denhaag-checkbox').click();
    cy.get('.denhaag-checkbox__input').should('not.be.checked');
  });

  it('can be disabled', () => {
    cy.mount(<Checkbox disabled></Checkbox>);

    cy.get('.denhaag-checkbox__input').should('have.attr', 'disabled');
  });

  it('can be operated by keyboard', () => {
    cy.mount(<Checkbox></Checkbox>);

    cy.get('.denhaag-checkbox__input').type(' ').should('be.checked');
  });

  it('a11y and snapshots', () => {
    cy.snapshots(
      <Checkbox
        onChange={() => {
          console.log('Checkbox action');
        }}
      />,
      {
        error: true,
        checked: true,
        disabled: true,
        uncheckedIcon: <HideIcon />,
      },
      {
        selector: '.denhaag-checkbox',
        states: {
          hover: 'denhaag-checkbox--hover',
          focus: 'denhaag-checkbox--focus',
        },
      },
      [{ id: 'label', logDisabled: true }],
    );
  });
});
