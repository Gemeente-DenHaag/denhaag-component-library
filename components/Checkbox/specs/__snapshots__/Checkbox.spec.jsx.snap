exports[`Checkbox tests > a11y and snapshots #0`] = `
<div id="wrapper" style="display: inline-block;">
  <div style="padding: 10px;">
    <span class="denhaag-checkbox"
      ><input type="checkbox" class="denhaag-checkbox__input"/><span
        class="denhaag-checkbox__icon"
        ><svg
          xmlns="http://www.w3.org/2000/svg"
          width="1em"
          height="1em"
          viewBox="0 0 24 24"
          class="denhaag-icon"
          focusable="false"
          aria-hidden="true"
          shape-rendering="auto"
        >
          <path
            d="M20.664 5.253a1 1 0 01.083 1.411l-10.666 12a1 1 0 01-1.495 0l-5.333-6a1 1 0 011.494-1.328l4.586 5.159 9.92-11.16a1 1 0 011.411-.082z"
            stroke="currentColor"
            stroke-width="1.75"
          ></path></svg></span
    ></span>
  </div>
  <div style="padding: 10px;">
    <span class="denhaag-checkbox denhaag-checkbox--error"
      ><input type="checkbox" class="denhaag-checkbox__input"/><span
        class="denhaag-checkbox__icon"
        ><svg
          xmlns="http://www.w3.org/2000/svg"
          width="1em"
          height="1em"
          viewBox="0 0 24 24"
          class="denhaag-icon"
          focusable="false"
          aria-hidden="true"
          shape-rendering="auto"
        >
          <path
            d="M20.664 5.253a1 1 0 01.083 1.411l-10.666 12a1 1 0 01-1.495 0l-5.333-6a1 1 0 011.494-1.328l4.586 5.159 9.92-11.16a1 1 0 011.411-.082z"
            stroke="currentColor"
            stroke-width="1.75"
          ></path></svg></span
    ></span>
  </div>
  <div style="padding: 10px;">
    <span class="denhaag-checkbox"
      ><input type="checkbox" class="denhaag-checkbox__input" checked=""/><span
        class="denhaag-checkbox__icon"
        ><svg
          xmlns="http://www.w3.org/2000/svg"
          width="1em"
          height="1em"
          viewBox="0 0 24 24"
          class="denhaag-icon"
          focusable="false"
          aria-hidden="true"
          shape-rendering="auto"
        >
          <path
            d="M20.664 5.253a1 1 0 01.083 1.411l-10.666 12a1 1 0 01-1.495 0l-5.333-6a1 1 0 011.494-1.328l4.586 5.159 9.92-11.16a1 1 0 011.411-.082z"
            stroke="currentColor"
            stroke-width="1.75"
          ></path></svg></span
    ></span>
  </div>
  <div style="padding: 10px;">
    <span class="denhaag-checkbox"
      ><input type="checkbox" class="denhaag-checkbox__input" disabled=""/><span
        class="denhaag-checkbox__icon"
        ><svg
          xmlns="http://www.w3.org/2000/svg"
          width="1em"
          height="1em"
          viewBox="0 0 24 24"
          class="denhaag-icon"
          focusable="false"
          aria-hidden="true"
          shape-rendering="auto"
        >
          <path
            d="M20.664 5.253a1 1 0 01.083 1.411l-10.666 12a1 1 0 01-1.495 0l-5.333-6a1 1 0 011.494-1.328l4.586 5.159 9.92-11.16a1 1 0 011.411-.082z"
            stroke="currentColor"
            stroke-width="1.75"
          ></path></svg></span
    ></span>
  </div>
  <div style="padding: 10px;">
    <span class="denhaag-checkbox"
      ><input
        type="checkbox"
        class="denhaag-checkbox__input"
        uncheckedicon="[object Object]"/><span class="denhaag-checkbox__icon"
        ><svg
          xmlns="http://www.w3.org/2000/svg"
          width="1em"
          height="1em"
          viewBox="0 0 24 24"
          class="denhaag-icon"
          focusable="false"
          aria-hidden="true"
          shape-rendering="auto"
        >
          <path
            d="M20.664 5.253a1 1 0 01.083 1.411l-10.666 12a1 1 0 01-1.495 0l-5.333-6a1 1 0 011.494-1.328l4.586 5.159 9.92-11.16a1 1 0 011.411-.082z"
            stroke="currentColor"
            stroke-width="1.75"
          ></path></svg></span
    ></span>
  </div>
</div>
`;
